HPCC-13495 Use correct default target directory for despray

wangkx opened the pull request:
The existing ESP code calls the RemoteFilename::setPath()
to set the target directory for despray. If a user does
not specify the target directory, the setPath() uses the
/var/lib/HPCCSystems/myesp as default target directory. In
this fix, If a user does not specify the target directory,
I use the target IP to retrieve the drop zone directory on
that IP and use the drop zone directory to be default
target directory.

Signed-off-by: wangkx kevin.wang@lexisnexis.com

wangkx committed:
HPCC-13495 Use correct default target directory for despray

The existing ESP code calls the RemoteFilename::setPath()
to set the target directory for despray. If a user does
not specify the target directory, the setPath() uses the
/var/lib/HPCCSystems/myesp as default target directory. In
this fix, If a user does not specify the target directory,
I use the target IP to retrieve the drop zone directory on
that IP and use the drop zone directory to be default
target directory.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@AttilaVamos could you please review?

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
@wangkx Needs rebasing for a clean merge

AttilaVamos commented:
Unfortunately it seems this changes didn't fix either DFUplus Despray nor ECL Playground problem. 
When this command executed:
dfuplus action=despray server=. srcname=.::xmlupload-2015-04-30_11-11 dstip=. dstfile=xmlupload-2015-04-30_11-11 overwrite=1

The desprayed file placed into the /var/lib/HPCCSystems/myesp directory instead of the dropzone.

The result is same for this ECL code in Playground:
import Std.File AS FileServices;

SrcIP := 'localhost';
File := '~xmlupload-2015-04-30_11-11';
DestFile := 'xmlupload-2015-04-30_11-11';

FileServices.Despray(   File,
                        SrcIp,
                        destinationPath := DestFile,
                        ALLOWOVERWRITE := true);

wangkx commented:
I do not know why the fix does not work for Attila.  This is my test command:

sudo ../../git2/my_cmake_ln/Debug/bin/dfuplus  action=despray server=. srcname=.::despray002 dstip=. dstfile=test55 overwrite=1

I got:
/var/lib/HPCCSystems/mydropzone/test55

wangkx commented:
@AttilaVamos could you please email me your environment.xml?

AttilaVamos commented:
I used the smoketest image which is build in a newly created environment from the pulled source. 
File sent.

wangkx commented:
My fix does not work for you because of the 'netAddress="."' in your environment.xml:
<Hardware>  <Computer computerType="linuxmachine"  domain="localdomain"      name="localhost"  netAddress="."/>
I will revise my fix.

wangkx commented:
@AttilaVamos please check again.

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

AttilaVamos commented:
It is fine for DFUplus command like:

dfuplus action=despray server=. srcname=.::xmlupload-2015-04-30_11-11 dstip=. dstfile=xmlupload-2015-04-30_11-11 overwrite=1

but still place desprayed file into myesp directory with this code in Playground:

import Std.File AS FileServices;
File := '~xmlupload-2015-04-30_11-11';
SrcIP := 'localhost';        // or '127.0.0.1'
DestFile := 'xmlupload-2015-04-30_11-11';

FileServices.Despray(   File,
                        SrcIp,
                        destinationPath := DestFile,
                        ALLOWOVERWRITE := true);

wangkx commented:
I guess it is because of the: SrcIP := 'localhost'. Let me check.

AttilaVamos commented:
Or SrcIP := '127.0.0.1'

wangkx commented:
Correct. The input could be '.', 'localhost', '127.0.0.1', or a real IP. And one of the options could be in environment.xml also.

wangkx commented:
@AttilaVamos please check again.

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

AttilaVamos commented:
@wangkx  It is fine now.

wangkx commented:
@richardkchapman could you please merge this PR?

richardkchapman commented:
@jakesmith Any dali efficiency concerns?

jakesmith commented:
I'm curious why we have a selectable drop zone drop list (that changes the 'IP Address:') and also the ability to ignore it and type in any IP ..? (It makes the checking/validation in this change more complicated)

It would be simpler if could only select a configured dropzone.. i.e. couldn't override the IP.
Perhaps shouldn't be able to enter an absolute path either.. only a relative one.

@richardkchapman - is entering custom IP + absolute path a 'feature' or could we insist have to pick a configured dropzone?

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2278,6 +2278,55 @@ bool CFileSprayEx::onReplicate(IEspContext &context, IEspReplicate &req, IEspRep
+    Owned<IPropertyTree> root = &env->getPTree();
+    VStringBuffer xPath("Hardware/Computer[@netAddress='%s']/@name", ip);
+    const char* computer = root->queryProp(xPath.str());
+    if (!computer || !*computer)
```
questioning whether this should ever happen really.
I realize it can at the moment, by virtue of the user being able to override the IP despite the drop list selection.
(See other comment).

wangkx replied:
BTW, the problem occurred when a user was using DFUPlus (not ECLWatch).

jakesmith replied:
Ok, well we could still say that has to find an entry, which it will do now for localhost variants.
But it's really a separate issue from the one being fixed here anyway.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2278,6 +2278,55 @@ bool CFileSprayEx::onReplicate(IEspContext &context, IEspReplicate &req, IEspRep
+        return NULL;
+
+    Owned<IPropertyTree> root = &env->getPTree();
+    VStringBuffer xPath("Hardware/Computer[@netAddress='%s']/@name", ip);
```
Isn't there something (or something close) to this functionality within the environment helpers?
Suspect there should be or it be extended to contain the functionality there.

wangkx replied:
I checked and haven't found one. If you want, I can move this one (getDropZoneDirByIP) there.

jakesmith replied:
I was referring to the Hardware/Computer info being looked up / extracted..

There are similar query method , that return interfaces in environment.*

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2278,6 +2278,55 @@ bool CFileSprayEx::onReplicate(IEspContext &context, IEspReplicate &req, IEspRep
+        if (!strieq("127.0.0.1", ip))
+            possibleIPs.append("127.0.0.1");
+        if (!strieq(localHostIP.str(), ip))
+            possibleIPs.append(localHostIP.str());
```
The conditions here are to avoid re-looking up one of these candidates again, since it will have done so for one on line 2293 ?
I wouldn't bother, the code will be a lot clearer if you just look up all the equivalents, until you find a match.

Again, this should probably be contained in environment.* , something like getMachineForLocalHost()

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2325,7 +2374,16 @@ bool CFileSprayEx::onDespray(IEspContext &context, IEspDespray &req, IEspDespray
+                StringBuffer buf;
+                getDropZoneDirByIP(destip, buf);
+                if (buf.length() && buf.charAt(buf.length()-1) != PATHSEPCHAR)
+                    buf.append(PATHSEPCHAR);
```
Could use:
if (buf.length())
    addPathSepChar(buf);

jakesmith commented:
@richardkchapman - not concerned about the efficiency here, but have some questions/comments

wangkx committed:
HPCC-13495 Add/use environment methods to get target dir

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith I made the changes based on your comments. Please review.

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -945,6 +980,9 @@ IConstDomainInfo * CLocalEnvironment::getDomain(const char * name) const
 void CLocalEnvironment::buildMachineCache() const
 {
+    StringBuffer localHostIP;
+    queryHostIP().getIpText(localHostIP);
```
minor: these lines may as well go in (if !machineCacheBuilt) block.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -966,12 +1004,42 @@ void CLocalEnvironment::buildMachineCache() const
                 x.append(name).append("\"]");
                 Owned<IConstEnvBase> cached = new CConstMachineInfo((CLocalEnvironment *) this, &it->query());
                 cache.setValue(x.str(), cached);
+                if (streq(".", name) || strieq("localhost", name) || streq("127.0.0.1", name) || streq(name, localHostIP.str()))
```
Strictly speaking anything matching 127.255.255.255 will point to the loopback device, so should really look for that.

Actually you should be using isLoopBack() from jsocket to test this.
Or if not a loopback IP, you should be using isLocal() to test if it's an interfface IP (as opposed to the test against localHostIP), since there can be >1 local interface.

jakesmith replied:
You don't need to test name == "." , as if "." cache entry already as added.

jakesmith replied:
In fact isLocal() will avoid need to explicitly test for "." as well, i.e. it is "." aware.

wangkx replied:
I am reading the code inside the isLocal(). Does it also handle "localhost"?

wangkx replied:
The isLocal() itself does not handle 'localhost'. But, the ipset() does. So, I just need to use "if (isLocal())" to replace this line.

jakesmith replied:
Yes.
For some reason I am not receiving email notifications from github for this PR (and perhaps others).

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1058,6 +1126,75 @@ IConstMachineInfo * CLocalEnvironment::getMachineByAddress(const char * name) co
+IConstMachineInfo * CLocalEnvironment::getMachineForLocalHost() const
+{
+    IConstEnvBase *cached = getCache("Hardware/Computer[@netAddress='.']");
+    if (!cached)
```
I don't you need most of the implementation here..

if you do:
```
buildMachineCache()
IConstEnvBase *cached = getCache("Hardware/Computer[@netAddress='.']");
```

Then buildMachineCache should have filled the "." cache entry.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1058,6 +1126,75 @@ IConstMachineInfo * CLocalEnvironment::getMachineByAddress(const char * name) co
+    VStringBuffer xpath("Software/DropZone[@name=\"%s\"]", name);
+    synchronized procedure(safeCache);
+    IConstEnvBase *cached = getCache(xpath.str());
+    if (!cached)
```
if the name exists as a valid DropZone, how can getCache have returned NULL, given buildDropZoneCache() has been called?

wangkx replied:
I wrote this method based on CLocalEnvironment::getMachine(). You are correct. If the name does not exist as a valid DropZone, it will return NULL even if I try one more time.

jakesmith commented on common/environment/environment.cpp:
```diff
@@ -1058,6 +1126,75 @@ IConstMachineInfo * CLocalEnvironment::getMachineByAddress(const char * name) co
+    VStringBuffer xpath("Software/DropZone[@computer='%s']", computer);
+    synchronized procedure(safeCache);
+    IConstEnvBase *cached = getCache(xpath.str());
+    if (!cached)
```
if the computer exists as a valid DropZone, how can getCache have returned NULL, given buildDropZoneCache() has been called?

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2278,6 +2278,39 @@ bool CFileSprayEx::onReplicate(IEspContext &context, IEspReplicate &req, IEspRep
+    {
+        StringBuffer localHostIP;
+        queryHostIP().getIpText(localHostIP);
+        if (!strieq(".", ip) && !strieq("localhost", ip) && !strieq("127.0.0.1", ip) && !strieq(ip, localHostIP.str()))
```
See other comments, but should test use isLocal() here too instead.

jakesmith commented:
@wangkx - Looks good generally, a few issues/comments.

wangkx committed:
HPCC-13495 Revise based on code review

1. Simplify code using isLocal()
2. Call getCache() since cache is built
3. Switch one ' to \" since it is the xpath being cached.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review my changes.

AttilaVamos commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

jakesmith commented:
Looks good, ready to merge.
@richardkchapman
